scapy/contrib/automotive/doip.py

```diff
@@ -0,0 +1,241 @@
+
+            matches = [(4, 1), (4, 2), (4, 3), (6, 5), (8, 7),
+                       (0x4002, 0x4001), (0x4004, 0x4003),
+                       (0x8001, 0x8001), (0x8003, 0x8001)]
+            if (self.payload_type, other.payload_type) in matches:
+                if self.payload_type == 0x8001:
+                    return self.payload.answers(other.payload)
+                return 1
+        return 0
+
+    def hashret(self):
+        # type: () -> bytes
+        if self.payload_type in [0x8001, 0x8002, 0x8003]:
+            return bytes(self)[:2] + struct.pack(
+                "H", self.target_address ^ self.source_address)
+        return bytes(self)[:2]
+
+    def post_build(self, pkt, pay):
+        # type: (bytes, bytes) -> bytes
+        """
+        This will set the Field 'payload_length' to the correct value.
+        """
+        if self.payload_length is None:
+            pkt = pkt[:4] + struct.pack("!I", len(pay) + len(pkt) - 8) + \
+                pkt[8:]
+        return pkt + pay
+
+    def extract_padding(self, s):
+        # type: (bytes) -> Tuple[bytes, Optional[bytes]]
+        if self.payload_type == 0x8001:
+            return s[:self.payload_length - 4], None
+        else:
+            return b"", None
+
+
+class DoIPSocket(StreamSocket):
+    def __init__(self, ip='127.0.0.1', port=13400, activate_routing=True,
+                 source_address=0xe80, target_address=0,
+                 activation_type=0):
+        # type: (str, int, bool, int, int, int) -> None
+        self.ip = ip
+        self.port = port
+        self.source_address = source_address
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        s.connect((self.ip, self.port))
+        StreamSocket.__init__(self, s, DoIP)
+
+        if activate_routing:
+            resp = self.sr1(
+                DoIP(payload_type=0x5, activation_type=activation_type,
+                     source_address=source_address),
+                verbose=False, timeout=1)
+            if resp and resp.payload_type == 0x6 and \
+                    resp.routing_activation_response == 0x10:
+                self.target_address = target_address or \
+                    resp.logical_address_doip_entity
+                print("Routing activation successful! "
+                      "Target address set to: 0x%x" % self.target_address)
+            else:
+                print("Routing activation failed! Response: %s" % repr(resp))
+
+
+class UDS_DoIPSocket(DoIPSocket):
+    def send(self, x):
+        # type: (Union[Packet, bytes]) -> int
+        if isinstance(x, UDS):
+            pkt = DoIP(payload_type=0x8001, source_address=self.source_address,
+                       target_address=self.target_address) / x
+        else:
+            pkt = x
+
+        try:
+            x.sent_time = time.time()  # type: ignore
```

Comment:
Why do you need this

Reply:
In case my `UDS_DoIPSocket` wraps an `UDS` packet into a `DoIP` packet, I want to populate the sent_time in this "upper" `UDS` packet. The underlaying function will populate `sent_time` in the `DoIP` packet, which is not known by the sender.